src/main/java/com/cleanroommc/groovyscript/compat/mods/prodigytech/AtomicReshaper.java

```diff
@@ -0,0 +1,179 @@
+        AtomicReshaperManager.INSTANCE.addRecipe(recipe);
+        addScripted(recipe);
+    }
+
+    @MethodDescription(priority = 2000, example = @Example(commented = true))
+    public void removeAll() {
+        AtomicReshaperManager.INSTANCE.getAllRecipes().forEach(this::addBackup);
+        AtomicReshaperManager.INSTANCE.removeAll();
+    }
+
+    @Property(property = "input", valid = @Comp("1"))
+    @Property(property = "output", valid = @Comp(value = "1", type = Comp.Type.GTE))
+    public static class RecipeBuilder extends AbstractRecipeBuilder<AtomicReshaperManager.AtomicReshaperRecipe> {
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE), defaultValue = "-1")
+        private int time = -1;
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int primordium;
+
+        private final List<Integer> outputWeights = new ArrayList<>();
+        private String errorMessage = "";
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder time(int time) {
+            this.time = time;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder primordium(int primordium) {
+            this.primordium = primordium;
+            return this;
+        }
+
+        public AtomicReshaper.RecipeBuilder output(ItemStack output) {
+            output(output, 1);
+            return this;
+        }
+
+        public AtomicReshaper.RecipeBuilder output(ItemStack... outputs) {
+            for (ItemStack output : outputs) {
+                output(output, 1);
+            }
+            return this;
+        }
+
+        public AtomicReshaper.RecipeBuilder output(Collection<ItemStack> outputs) {
+            for (ItemStack output : outputs) {
+                output(output, 1);
+            }
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder output(ItemStack output, int weight) {
+            if (weight <= 0 && errorMessage.isEmpty()) {
+                errorMessage = String.format("ItemStack %s has invalid weight %d, expected >= 1", output, weight);
+            } else {
+                this.output.add(output);
+                outputWeights.add(weight);
+            }
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding ProdigyTech Atomic Reshaper Recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            validateItems(msg, 1, 1, 1, Integer.MAX_VALUE);
+            validateFluids(msg);
+
+            // I think this check is not possible to fail at all but still, adding it for consistency
+            msg.add(output.size() != outputWeights.size(), "Outputs and output weights must be the same size!");
+
+            msg.add(!errorMessage.isEmpty(), "Weighted outputs error: {}", errorMessage);

```

Comment:
```suggestion
            msg.add(outputWeights.stream().anyMatch(x -> x <= 0), "all weighted outputs must be greater than 0, yet they were {}", outputWeights);
```

Reply:
oh thats way better, that was kinda nasty to do with the old `Object[]` version